src/Streamly/Internal/FileSystem/Handle.hs

```diff
@@ -481,6 +491,65 @@ write = writeWithBufferOf defaultChunkSize
 write2 :: MonadIO m => Fold2 m Handle Word8 ()
 write2 = writeWithBufferOf2 defaultChunkSize
 
+
+toChunksWithBufferOfRange
+    :: (IsStream t, MonadIO m) 
+    => Int 
+    -> IORef Int 
+    -> Handle 
+    -> t m (Array Word8)
+toChunksWithBufferOfRange size lenRef h = D.fromStreamD (D.Stream step ())
+  where
+    {-# INLINE_LATE step #-}
+    step _ _ = do
+        remain <- liftIO $ readIORef lenRef
+        arr <- liftIO $ readArrayUpto (min remain size) h
+        liftIO $ writeIORef lenRef (remain - (A.length arr)) 
+        return $
+            case A.length arr of
+                0 -> D.Stop
+                _ -> D.Yield arr ()    
+
+{-# INLINE readSliceWith #-}
+readSliceWith :: (IsStream t, MonadIO (t m), MonadIO m)
+    => Int -> Handle -> Int -> Int -> t m (Array Word8)
+readSliceWith chunkSize h pos len = do    
+    lenRef <- liftIO $ newIORef len
+    liftIO $ hSeek h AbsoluteSeek $ fromIntegral pos
+    toChunksWithBufferOfRange chunkSize lenRef h
+
+{-# INLINABLE writeArrayWith #-}
+writeArrayWith :: Storable a => Handle -> Array a -> Int -> IO ()
+writeArrayWith _ arr _ | A.length arr == 0 = return ()
+writeArrayWith h Array{..} len = withForeignPtr aStart $ \p -> hPutBuf h p len  
+
+{-# INLINE fromChunksWith #-}
+fromChunksWith :: (MonadIO m, Storable a)
+    => Handle -> IORef Int -> SerialT m (Array a) -> m ()
+fromChunksWith h lenRef = do    
+    S.mapM_ (\arr -> do 
```

Comment:
You can probably take this function out into a where clause and rewrite this as `S.mapM_ func`